src/ir/mod.rs

```diff
@@ -447,17 +447,74 @@ impl PolarizedTraitRef {
     }
 }
 
+/// "Basic" where clauses which have a WF/FromEnv version of themselves.
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum WhereClauseAtom {
+    Implemented(TraitRef),
+    ProjectionEq(ProjectionEq),
+}
+
 /// A "domain goal" is a goal that is directly about Rust, rather than a pure
 /// logical statement. As much as possible, the Chalk solver should avoid
 /// decomposing this enum, and instead treat its values opaquely.
 #[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum DomainGoal {
-    Implemented(TraitRef),
-    ProjectionEq(ProjectionEq),
+    Holds(WhereClauseAtom),
+
+    /// A predicate which is true is some trait ref is well-formed.
+    /// For example, given the following trait definitions:
+    /// 
+    /// ```notrust
+    /// trait Clone { ... }
+    /// trait Copy where Self: Clone { ... }
+    /// ```
+    ///
+    /// then we have the following rule:
+    /// `WellFormed(?Self: Copy) :- ?Self: Copy, WellFormed(?Self: Clone)`.
+    WellFormed(WhereClauseAtom),
+
+    /// A predicate which enables deriving everything which should be true if we *know* that
+    /// some trait ref is well-formed. For example given the above trait definitions, we can use
+    /// `FromEnv(T: Copy)` to derive that `T: Clone`, like in:
+    ///
+    /// ```notrust
+    /// forall<T> {
+    ///     if (FromEnv(T: Copy)) {
+    ///         T: Clone
+    ///     }
+    /// }
+    /// ```
+    FromEnv(WhereClauseAtom),
+
+
     Normalize(Normalize),
     UnselectedNormalize(UnselectedNormalize),
-    WellFormed(WellFormed),
-    FromEnv(FromEnv),
+
+    /// A predicate which is true is some type is well-formed.
+    /// For example, given the following type definition:
+    ///
+    /// ```notrust
+    /// struct Set<K> where K: Hash {
+    ///     ...
+    /// }
+    /// ```
+    ///
+    /// then we have the following rule: `WellFormedTy(Set<K>) :- Implemented(K: Hash)`.
+    WellFormedTy(Ty),
```

Comment:
Why is this .. oh, it's not a `WhereClauseAtom` because `Holds(Ty)` makes no sense? I wonder if it'd be worth adding it just for symmetry. We just happen to have no rules that establish `HoldsTy` =)

Reply:
Yeah I don't know, actually `WF(ty)` is a bit different from `WF(Trait)`: it is not co-inductive and it does need to "recursively" prove bounds (just the bounds directly expressed on the type). I guess: let's keep it separated for the moment.